### DISCUSSION TITLE
how to inject queryKey explicitly into atomsWithQuery

### DISCUSSION THREAD

**Discussion by dante01yoon** [pmndrs/jotai#1578] [Q&A] [answered]
Hi, 

in react-query, I normally write custom hook to inject different keys in different `useQuery` 

```tsx
const useCustomQuery = (key: string) => {
 const queryInfo = useQuery({
   queryKey: [key]
   ... 
 });

 return {
   ...queryInfo,
   data: queryInfo.data
 }
}
```

how can I accomplish this with atomsWithQuery?

**Comment by dai-shi** [comment, marked as answer]
If `key` is from props, so you can create atoms only in hooks, not globally. This is not specific to atomsWithQuery.

```js
const useCustomAtomsWithQuery = (key) => {
  return useMemo(() => atomWithQuery(...), [key])
}
```

**Reply by dante01yoon** [reply]
I see. thank you!

**Comment by dante01yoon** [comment]
@dai-shi  I think there is some misleading in [docs about atomWithQuery](https://jotai.org/docs/integrations/query).

In the doc return type of atomWithQuery is iterable, but real response type is not. 
type error occurs when try get response via array destructing. 

```ts
 // Type 'WritableAtom<string[], AtomWithQueryAction, void>' must have a '[Symbol.iterator]()' method that returns an iterator.ts(2488)
const [stringArrayAtom] = atomWithQuery((get) => {
  ... 
})
```
I should have changed this into below 
```ts
const stringArrayAtom = atomWithQuery((get) => {
  ... 
})
```

**Reply by dai-shi** [reply]
Docs are already migrated to a new library.
```js
import { atomsWithQuery } from 'jotai-tanstack-query'
```

You can still use `import { atomWithQuery } from 'jotai/query'` now, but it will be deprecated soon.

**Reply by dante01yoon** [reply]
I see! thank you
